### DISCUSSION TITLE
Custom Formatter Error

### DISCUSSION THREAD

**Discussion by snarkipus** [eclipse-langium/langium#624] [Q&A] [answered]
Tag: @msujew 

Example Project: [Entities DSL Repo](https://github.com/snarkipus/langium-entities)

I'm attempting to implement a custom formatter by extending the AbstractFormatter as shown in the Domain example. The formatter should insert two newlines after each `entity` thereby separating entities by an empty line. Additionally, there should be a single empty line after the last `entity`.

Example Text:
```js
entity MyFirstEntity {
    string s;
    int[10] a;
}

entity MySecondEntity extends MyFirstEntity {
    MyFirstEntity e;
}
```
[entities-formatter.ts](https://github.com/snarkipus/langium-entities/blob/4848d17091ef11924e9583677be4f84edcbc8548/src/language-server/entities-formatter.ts#L1-L20)
```ts
import { AbstractFormatter, Formatting } from 'langium';
import * as ast from './generated/ast';

export class EntitiesFormatter extends AbstractFormatter {

    protected format(model: ast.Model): void {

        const lastEntity = model.entities[model.entities.length - 1];

        model.entities.forEach(entity => {
            const formatter = this.getNodeFormatter(entity);
            const node = formatter.node(entity);
            if (entity === lastEntity) {                
                node.append(Formatting.newLine());
            } else {
                node.append(Formatting.newLines(2));
            }
        });
    }
}
```

Invoking the format command in VSCode using `Shift+Alt+F` produces the following:
```terminal
Debugger listening on ws://127.0.0.1:6009/e658526d-e20a-4340-a421-88b4a3dc7add
For help, see: https://nodejs.org/en/docs/inspector
[Error - 12:45:26 PM] Request textDocument/formatting failed.
  Message: Request textDocument/formatting failed with message: Cannot read properties of undefined (reading 'length')
  Code: -32603 
```

While intuitively I assume it's complaining about trying to read the array length, I'm not sure how I would implement it differently.

**Comment by snarkipus** [comment, marked as answer]
Tag: @msujew 

Disregard - I'll leave this here in case anyone else is curious. I'm assuming that since we don't have dynamic dispatch (Xtend) that means that we have to call `format()` on all of the nodes of the AST and select which nodes to apply formatting to using the utility `isSomeType()` function. Please feel free to clarify if I'm off the mark.

```ts
    protected format(node: AstNode): void {

        if (ast.isModel(node)) {
            const lastEntity = node.entities[node.entities.length - 1];

            node.entities.forEach(entity => {
                const formatter = this.getNodeFormatter(entity);
                const node = formatter.node(entity);
                if (entity === lastEntity) {                
                    node.append(Formatting.newLine());
                } else {
                    node.append(Formatting.newLines(5));
                }
            });
        }
    }
```

**Reply by msujew** [reply]
Hey, great that you were able to resolve the issue :) You're right, that the `format` function is a general entry point for formatting, and the abstract formatter passes every node in the document into it. I'm positively surprised that it works for you, since I've expected more issues with this due to our lack of automated tests for it.

Also, I've been planning to document the formatting feature (or rather: write a guide on it), but didn't have the time yet. I'm currently trying to push the documentation forward, so there probably will be some progress in the coming weeks.

**Comment by snarkipus** [comment]
This seems to be _mostly_ working.

For some add'l context - this is what I'm trying to "port" for lack of better words:
* [Xtend Formatting Class](https://github.com/PacktPublishing/Implementing-Domain-Specific-Languages-with-Xtext-and-Xtend/blob/c2060cabe814d316802ad326d87655db255cfc32/org.example.entities.parent/org.example.entities/src/org/example/entities/formatting2/EntitiesFormatter.xtend#L16-L57)
```java
class EntitiesFormatter extends AbstractFormatter2 {

	@Inject extension EntitiesGrammarAccess

	def dispatch void format(Model model, extension IFormattableDocument document) {
		val lastEntity = model.entities.last
		for (entity : model.entities) {
			entity.format
			if (entity === lastEntity)
				entity.append[setNewLines(1)]
			else
				entity.append[setNewLines(2)]
		}
	}

	def dispatch void format(Entity entity, extension IFormattableDocument document) {
		entity.regionFor.keyword("extends").surround[oneSpace]
		entity.regionFor.feature(ENTITY__NAME).surround[oneSpace]
		entity.regionFor.feature(ENTITY__SUPER_TYPE).surround[oneSpace]

		val open = entity.regionFor.keyword(entityAccess.leftCurlyBracketKeyword_3)
		val close = entity.regionFor.keyword("}")
		open.append[newLine]
		interior(open, close)[indent]

		for (attribute : entity.attributes) {
			attribute.type.format(document)
			attribute.regionFor.keyword(";").surround[noSpace]
			attribute.append[setNewLines(1, 1, 2)]
		}
	}

	def dispatch void format(AttributeType attributeType, extension IFormattableDocument document) {
		if (!attributeType.array) {
			attributeType.elementType.append[oneSpace]
		} else {
			attributeType.regionFor.keyword("[").surround[noSpace]
			attributeType.regionFor.keyword("]").prepend[noSpace].append[oneSpace]
		}
	}

}
```
* [Langium Custom Formatter](https://github.com/snarkipus/langium-entities/blob/db1d1e83c93c574743adcc193dc1a840c91b64f8/src/language-server/entities-formatter.ts#L4-L48)
```ts
export class EntitiesFormatter extends AbstractFormatter {

    protected format(node: AstNode): void {

        if (ast.isModel(node)) {
            const lastEntity = node.entities[node.entities.length - 1];

            node.entities.forEach(entity => {
                const formatter = this.getNodeFormatter(entity);
                const node = formatter.node(entity);
                if (entity === lastEntity) {                
                    node.append(Formatting.newLine());
                } else {
                    node.append(Formatting.newLines(2));
                }
            });
        } else if (ast.isEntity(node)) {
            const formatter = this.getNodeFormatter(node);
            formatter.keyword("extends").surround(Formatting.oneSpace());
            // entity name
            // entity superType
            
            const bracesOpen = formatter.keyword("{");
            const bracesClose = formatter.keyword("}");
            bracesOpen.append(Formatting.newLine());
            formatter.interior(bracesOpen, bracesClose).prepend(Formatting.indent());

            node.attributes.forEach(attribute => {
                const formatter = this.getNodeFormatter(attribute.type);
                const node = formatter.node(attribute);
                formatter.keyword(";").surround(Formatting.noSpace());
                node.append(Formatting.newLine());
            });
        } else if (ast.isAttributeType(node)) {
            const formatter = this.getNodeFormatter(node);
            if (!node.array) {
                const elem = formatter.node(node.entity);
                elem.append(Formatting.oneSpace());
            } else {
                formatter.keyword("[").surround(Formatting.noSpace());
                formatter.keyword("]").prepend(Formatting.noSpace()).append(Formatting.oneSpace());
            }
        }
    }
}
```

**Reply by msujew** [reply]
> This seems to be mostly working.

Any particular issues that you're noticing? The formatting stuff is missing automated tests, so there might be a few issues in there. The API also might not be the most intuitive...

**Reply by snarkipus** [reply]
Well, it's definitely wonky - but I'm not 100% sure that it's not my implementation. For instance, the formatting will oscillate on repeated format invocations (i.e., right now it looks better if I format it twice).

Since there is a single entry point through the `format` function, I suspect that some thought needs to be placed regarding precedence and the else-if fall-through sequence.  I'd have to think about it, but it would seem like if it is formatting _all_ of the nodes, then you'd have to be strategic with the order ... most likely from bottom to top (or inside to outside).

I believe it should be repeatable if you were inclined to pull that example repo down and build/run it. I'll play with it some more ... it seemed like some of the prepend/append formatting for newlines and indents was unpredictable. I'll note anything I find and open an issue if I can see it's clearly borked.

Luckily the next section in this book is on automated testing for the formatting. :)

... or not. Just discovered the utility of the helper classes you guys create. Figuring out how to parse text into AstNodes and emulate a LSP call to the service is a bit beyond my skillset (if that's even how it's done).

**Reply by msujew** [reply]
> ... or not. Just discovered the utility of the helper classes you guys create. Figuring out how to parse text into AstNodes and emulate a LSP call to the service is a bit beyond my skillset (if that's even how it's done).

Luckily, that's not even necessary. You can relatively easily create your own test method for this: You just need the `parseHelper`, the `Formatter` and then combine the resulting `LangiumDocument.textDocument` and `TextEdit[]` in the `TextDocument.applyEdits` function. I actually just found out that this function exists...

**Reply by msujew** [reply]
See https://github.com/langium/langium/pull/626

**Comment by snarkipus** [comment]
Tag: @msujew 

I'll just tag this here ... working through some example parser unit tests and what seems like a straightforward object equality check results in a heap error. It's not completely obvious to me why this explodes ...

[test repo](https://github.com/snarkipus/langium-expressions)

### Grammar:
```
grammar Expressions

entry ExpressionsModel:
    elements+=AbstractElement*;

AbstractElement:
    Variable | EvalExpression;

Variable:
    'var' name=ID '=' expression=Expression;

EvalExpression:
    'eval' expression=Expression;

Expression:
      { infer IntConstant    } value=INT 
    | { infer StringConstant } value=STRING 
    | { infer BoolConstant   } value=('true' | 'false')
    | { infer VariableRef    } variable=[Variable];

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal INT returns number: /[0-9]+/;
terminal STRING: /"[^"]*"|'[^']*'/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

### Unit Test ( ... this goes boom ... ):
```ts
test('Parse: VariableReference', async () => {
    const text=`
    var i = 10
    eval i`;
    const parse = await helper(text);
    const elements = parse.parseResult.value.elements;
    expect((elements[elements.length - 1].expression as VariableRef).variable).toEqual(elements[0]);
}); 
```

Note: As a work around (which I think accomplishes the intent of testing the reference), I used the following, which seemed to work.
```ts
expect((elements[elements.length - 1].expression as VariableRef).variable.ref?.name).toBe('i')
```

**Reply by msujew** [reply]
I believe the issue is that jest performs a naive deep quality comparison with `toEqual`. However, all AST and CST nodes are cyclic (since they have a bidirectional parent <-> child relationship). So a naive comparison will likely run into a stack overflow. You should use `toBe` instead, if you want to assert object identity.

**Reply by snarkipus** [reply]
Tag: @msujew 

Thanks so much for taking all this time to respond! Just start signing me up for easy bugs to fix or documentation so I can repay you guys for all of the education.

So that makes sense (I think). Just to make sure I'm following ... 

These unit tests are all originally written in Xtend ... so in that case they are most likely evaluating true object equality. I'm guessing that Xtext must be smart enough to resolve the declared variable token and the variable reference token to the same AST object (or something along those lines ... sounds close enough).

Using Langium, they are separate, bi-directionally linked elements ... so true equality (naive?) will never resolve. In that case, testing for correct reference is more appropriate (i.e. the workaround-actually-most-likely-correct test I wrote).

On the topic of porting Xtend tests to Jest ... which is more difficult than I anticipated ...

Any thoughts on this? I'm trying to replicate an associativity test for the same test repo I linked above. I apologize in advance for the length of this, it really didn't start off this long ... it ended up being a stream of consciousness. 

### Grammar:
```
grammar Expressions

entry ExpressionsModel:
    elements+=AbstractElement*;

AbstractElement:
    Variable | EvalExpression;

Variable:
    'var' name=ID '=' expression=Expression;

Expression:
    Atomic ({ infer Plus.left=current } '+' right=Atomic)*;

Atomic infers Expression:
      '(' Expression ')'
    | { infer IntConstant    } value=INT 
    | { infer StringConstant } value=STRING 
    | { infer BoolConstant   } value=('true' | 'false')
    | { infer VariableRef    } variable=[Variable];
```
As best I can tell, the Xtend unit test (shown below) is doing the following:
* Calls the `assertRepr()` function, which builds the pass/fail test case
  * input string (CharSequence in java ...):  `(((10 + 5) + 1) + 2)`
  * expected output string: `10 + 5 + 1 + 2`
* `assertRepr()`:
  *  Parses `eval (((10 + 5) + 1) + 2)`  (CST -> AST)
  * Calls the `stringRepr()` function which builds a string representation of the parsed AST
    * input Expression:  `elements.last.expression`
    * output string: _should_ resolve to `INT+INT+INT+INT` after recursion?
  * Performs an object equality test (naive probably?) between the specified `expected` string and generated output string

### Xtend Associativity Test (partial):
```java
	@Test
	def void testPlus() {
		"10 + 5 + 1 + 2".assertRepr("(((10 + 5) + 1) + 2)")
	}

	def private String stringRepr(Expression e) {
		switch (e) {
			Plus:		'''(«e.left.stringRepr» + «e.right.stringRepr»)'''
			IntConstant: 	'''«e.value»'''
			StringConstant:	'''«e.value»'''
			BoolConstant:	'''«e.value»'''
			VariableRef:	'''«e.variable.name»'''
		}.toString
	}
	
	def private assertRepr(CharSequence input, CharSequence expected) {
		("eval " + input).parse => [
			expected.assertEquals(elements.last.expression.stringRepr)
		]
	}
```
So, I took a shot at doing this using Langium:
### Langium Associativity Test
```ts
import { parseHelper } from 'langium/test';
import * as ast from '../language-server/generated/ast';
import { createExpressionsServices } from '../language-server/expressions-module';

const services = createExpressionsServices().Expressions;
const helper = parseHelper<ast.ExpressionsModel>(services);

describe('Expressions Parser', () => {
    it('should have associative addition', async () => {
        const input = `(((10 + 5) + 1) + 2)`;
        const expected = `10 + 5 + 1 + 2`;
        expect(assertRepr(input)).resolves.toBe(expected);
    });

    function stringRepr(e: ast.Expression): string  {
        switch (e.$type) {
            case ast.Plus:           return `(${stringRepr((e as ast.Plus).left)} + ${stringRepr((e as ast.Plus).right)})`;
            case ast.IntConstant:    return `${(e as ast.IntConstant).value}`; 
            case ast.StringConstant: return `${(e as ast.StringConstant).value}`;
            case ast.BoolConstant:   return `${(e as ast.BoolConstant).value}`;
            case ast.VariableRef:    return `${(e as ast.VariableRef).variable.ref?.name}`;
            default:                 throw new Error(`Unknown expression type: ${e.$type}`);
        }
    }

    const assertRepr = async (input: string) => {
        try {
            const parse = await helper(input);
            const elements = parse.parseResult.value.elements;
            console.log(parse.parseResult);
            return stringRepr(parse.parseResult.value.elements[elements.length - 1].expression);
        } catch(error) {
            return error;
        }
    }
});
```
### Result
```terminal
  console.log
    <ref *2> {
      value: <ref *1> {
        '$type': 'ExpressionsModel',
        '$cstNode': RootCstNodeImpl {
          children: [CstNodeContainer],
          _text: '(((10 + 5) + 1) + 2)',
          _element: [Circular *1]
        },
        elements: [],
        '$document': {
          parseResult: [Circular *2],
          uri: [e],
          state: 5,
          references: [],
          textDocument: [Getter],
          precomputedScopes: [MultiMap]
        }
      },
      lexerErrors: [],
      parserErrors: [
        MismatchedTokenException: Expecting token of type 'EOF' but found `(`.
            at generateErrorMessage (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recoverable.ts:113:21)
            at ChevrotainWrapper.Object.<anonymous>.Recoverable.tryInRepetitionRecovery (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recoverable.ts:126:9)
            at ChevrotainWrapper.attemptInRepetitionRecovery (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recoverable.ts:450:10)
            at ChevrotainWrapper.Object.<anonymous>.RecognizerEngine.manyInternalLogic (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recognizer_engine.ts:514:10)
            at ChevrotainWrapper.Object.<anonymous>.RecognizerEngine.manyInternal (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recognizer_engine.ts:481:17)
            at ChevrotainWrapper.Object.<anonymous>.RecognizerApi.many (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/chevrotain/src/parse/parser/traits/recognizer_api.ts:76:17)
            at ChevrotainWrapper.wrapMany (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/langium/src/parser/langium-parser.ts:459:14)
            at LangiumParser.many (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/langium/src/parser/langium-parser.ts:169:22)
            at idx (/home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/langium/src/parser/langium-parser-builder.ts:381:37)
            at /home/snarkipus/Projects/langium/xtext-book/expressions/node_modules/langium/src/parser/langium-parser-builder.ts:107:9 {
          token: [Object],
          resyncedTokens: [Array],
          previousToken: [Object],
          context: [Object]
        }
      ]
    }

      at src/test/expressions-parser-test.test.ts:61:21


 RUNS  src/test/expressions-parser-test.test.ts
Waiting for the debugger to disconnect...
node:internal/process/promises:265
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Error: expect(received).resolves.toBe(expected) // Object.is equality

Expected: "10 + 5 + 1 + 2"
Received: [TypeError: Cannot read properties of undefined (reading 'expression')]".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```
Ok ... so `parse()` didn't produce any elements as indicated by the log and confirmed by the `TypeError: Cannot read properties of undefined (reading 'expression')` statement. This must be because Chevrotain hates both me and my parentheses as indicated by `MismatchedTokenException: Expecting token of type 'EOF' but found `(`.`

At this point, I'm out on a far enough ledge to see if I'm just doing something silly.

**Reply by snarkipus** [reply]
@msujew Disregard - missing `eval` statement and my input & output were reversed. I'll leave the final working code here for posterity.

### Associativity Test Code (Final)
```ts
describe('Expressions Parser', () => {
    it('should have associative addition', async () => {
        const input    = `10 + 5 + 1 + 2`;
        const expected = `(((10 + 5) + 1) + 2)`;
        expect(await assertRepr(input)).toBe(expected);
    });

    function stringRepr(e: ast.Expression): string  {
        switch (e.$type) {
            case ast.Plus:           return `(${stringRepr((e as ast.Plus).left)} + ${stringRepr((e as ast.Plus).right)})`;
            case ast.IntConstant:    return `${(e as ast.IntConstant).value}`; 
            case ast.StringConstant: return `${(e as ast.StringConstant).value}`;
            case ast.BoolConstant:   return `${(e as ast.BoolConstant).value}`;
            case ast.VariableRef:    return `${(e as ast.VariableRef).variable.ref?.name}`;
            default:                 throw new Error(`Unknown expression type: ${e.$type}`);
        }
    }

    const assertRepr = async (input: string) => {
            const parse = await helper('eval ' + input);
            const elements = parse.parseResult.value.elements;
            return stringRepr(parse.parseResult.value.elements[elements.length - 1].expression);
    }
});
```
